File: app/src/main/java/com/rokudo/xpense/models/Transaction.java
package com.rokudo.xpense.models;

import java.io.Serializable;
import java.text.DecimalFormat;
import java.util.Date;
import java.util.Objects;

public class Transaction implements Serializable {
    private String id;
    private String walletId;
    private String type;
    private String category;
    private String user_id;
    private String userName;
    private String title;
    private Double amount;
    private String currency;
    private Date date;
    private String picUrl;


    public static final String INCOME_TYPE = "Income";
    public static final String EXPENSE_TYPE = "Expense";


    public Transaction() {
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getWalletId() {
        return walletId;
    }

    public void setWalletId(String walletId) {
        this.walletId = walletId;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getUser_id() {
        return user_id;
    }

    public void setUser_id(String user_id) {
        this.user_id = user_id;
    }

    public String getUserName() {
        return userName;
    }

    public void setUserName(String userName) {
        this.userName = userName;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public Double getAmount() {
        DecimalFormat decimalFormat = new DecimalFormat("0.00");
        return Double.parseDouble(decimalFormat.format(amount));
    }

    public void setAmount(Double amount) {
        this.amount = amount;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public Date getDate() {
        return date;
    }

    public void setDate(Date date) {
        this.date = date;
    }

    public String getPicUrl() {
        return picUrl;
    }

    public void setPicUrl(String picUrl) {
        this.picUrl = picUrl;
    }

    @Override
    public String toString() {
        return "Transaction{" +
                "id='" + id + '\'' +
                "walletId='" + walletId + '\'' +
                ", type='" + type + '\'' +
                ", category='" + category + '\'' +
                ", user_id='" + user_id + '\'' +
                ", userName='" + userName + '\'' +
                ", title='" + title + '\'' +
                ", amount=" + amount +
                ", currency='" + currency + '\'' +
                ", date=" + date +
                ", picUrl='" + picUrl + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Transaction that = (Transaction) o;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: app/src/main/java/com/rokudo/xpense/models/Transaction.kt
package com.rokudo.xpense.models

import java.io.Serializable
import java.text.DecimalFormat
import java.util.Date

data class Transaction(
    var id: String? = null,
    var walletId: String? = null,
    var type: String? = null,
    var category: String? = null,
    var user_id: String? = null,
    var userName: String? = null,
    var title: String? = null,
    var amount: Double? = null,
    var currency: String? = null,
    var date: Date? = null,
    var picUrl: String? = null,
    var isCashTransaction: Boolean? = null,
    var alreadyAdded: Boolean? = null
) : Serializable {

    companion object {
        const val INCOME_TYPE = "Income"
        const val EXPENSE_TYPE = "Expense"
        const val TRANSFER_TYPE = "Transfer"
    }

    fun getFormattedAmount(): Double {
        val df = DecimalFormat("0.00")
        return df.format(amount ?: 0.0).toDouble()
    }

    val dateLong: Long
        get() = date?.time ?: 0L

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Transaction) return false
        return id == other.id
    }

    override fun hashCode(): Int = id?.hashCode() ?: 0
}
